gen.ts

```diff
@@ -0,0 +1,217 @@
+			}
+		}
+
+		const buildersString = builders.join("\n");
+		const versionsString = versions.join("\n");
+		const latestAliasesString = latestAliases.join("\n");
+
+		return `
+			{ callPackage, fetchurl }:
+
+			let
+				build = { ${buildersString} };
+
+				versions = { ${versionsString} };
+
+				latest-aliases = { ${latestAliasesString} };
+			in versions // latest-aliases
+		`;
+	}
+}
+
+async function nodejs(): Promise<string> {
+	const NodeIndexJson = Type.Array(
+		Type.Object({
+			version: Type.String(),
+			date: Type.String(),
+			files: Type.Array(Type.String()),
+			v8: Type.String(),
+			lts: Type.Union([Type.String(), Type.Boolean()]),
+			security: Type.Boolean(),
+
+			// the earliest versions don't have these fields
+			npm: Type.Optional(Type.String()),
+			uv: Type.Optional(Type.String()),
+			zlib: Type.Optional(Type.String()),
+			openssl: Type.Optional(Type.String()),
+			modules: Type.Optional(Type.String()),
+		}),
+	);
+
+	// step 1: fetch the node.js version index
+	const nodejsIndexURL = "https://nodejs.org/download/release/index.json";
+	const nodejsIndexResp = await fetch(nodejsIndexURL);
+	if (!nodejsIndexResp.ok) {
+		throw new Error(
+			`Failed to fetch node.js release index from ${nodejsIndexURL}: ${nodejsIndexResp.statusText}`,
+		);
+	}
+
+	const nodejsIndexJson = await nodejsIndexResp.json();
+	if (!Value.Check(NodeIndexJson, nodejsIndexJson)) {
+		const errors = [...Value.Errors(NodeIndexJson, nodejsIndexJson)];
+		throw new Error(
+			`malformed response for nodejs index from ${nodejsIndexURL}`,
+			{
+				cause: errors,
+			},
+		);
+	}
+
+	// step 2: parse the versions, filter out the ones we aren't handling, and sort (to ensure consistency)
+	const versions = nodejsIndexJson
+		.map((release) => {
+			return {
+				versionString: release.version.substring(1),
+				version: release.version
+					.substring(1)
+					.split(".")
+					.map((vPart) => parseInt(vPart, 10)) as [number, number, number],
+			};
+		})
+		.filter(({ version }) => version[0] >= 21)
+		.sort(
+			(
+				{ version: [leftMajor, leftMinor, leftPatch] },
+				{ version: [rightMajor, rightMinor, rightPatch] },
+			) => {
+				const majorDiff = leftMajor - rightMajor;
+				if (majorDiff !== 0) {
+					return majorDiff;
+				}
+
+				const minorDiff = leftMinor - rightMinor;
+				if (minorDiff !== 0) {
+					return minorDiff;
+				}
+
+				return leftPatch - rightPatch;
+			},
+		);
+
+	// step 3: insert each version into the version map
+	const versionMap = new MajorVersionMap();
+	// match `deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef	node-v99.99.99.tar.xz`
+	const sourceHashLineReg =
+		/^([a-fA-F0-9]{64})\s+node-v(\d{1,2}\.\d{1,2}\.\d{1,2})\.tar\.xz$/;
+	for (const { version, versionString } of versions) {
+		// step 3a: get the SHASUMS256.txt file of the target release
+		const hashFileURL = `https://nodejs.org/download/release/v${versionString}/SHASUMS256.txt`;
+		const hashFileResp = await fetch(hashFileURL);
```

Comment:
Did you try using nix-prefetch-url?

Reply:
yes but i'd rather rely on the `SHASUMS256.txt` file since i'm treating the nodejs.org registry as the source of truth